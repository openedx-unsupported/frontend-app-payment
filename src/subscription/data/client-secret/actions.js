import { createRoutine } from 'redux-saga-routines';

// Created with redux-saga-routines
export const fetchSubscriptionClientSecret = createRoutine('FETCH_SUBSCRIPTION_CLIENT_SECRET');

export const SUBSCRIPTION_CLIENT_SECRET_PROCESSING = 'SUBSCRIPTION_CLIENT_SECRET_PROCESSING';
export const SUBSCRIPTION_CLIENT_SECRET_DATA_RECEIVED = 'SUBSCRIPTION_CLIENT_SECRET_DATA_RECEIVED';

export const clientSecretProcessing = isProcessing => ({
  type: SUBSCRIPTION_CLIENT_SECRET_PROCESSING,
  payload: isProcessing,
});

export const clientSecretDataReceived = clientSecret => ({
  type: SUBSCRIPTION_CLIENT_SECRET_DATA_RECEIVED,
  // TODO: refactor this to get the clientSecretID when fetching from subs BE
  payload: clientSecret && clientSecret.capture_context ? clientSecret.capture_context.key_id : null,
});
