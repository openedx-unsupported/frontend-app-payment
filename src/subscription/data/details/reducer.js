import {
  SUBSCRIPTION_DETAILS_RECEIVED,
  SUBSCRIPTION_DETAILS_PROCESSING,
  fetchSubscriptionDetails,
  submitPayment,
} from './actions';

const subscriptionDetailsInitialState = {
  loading: true,
  loaded: false,
  submitting: false,
  redirect: false,
  enableStripePaymentProcessor: true,
  isSubscriptionDetailsProcessing: false,
  products: [],
  // TODO: remove these once fetched from DB
  currency: 'USD',
  programTitle: 'Blockchain Fundamentals',
  certificateType: 'verified',
  organization: 'University of California, Berkeley',
  price: 49,
  paymentMethod: 'stripe',
};

export const subscriptionDetailsReducer = (state = subscriptionDetailsInitialState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case fetchSubscriptionDetails.TRIGGER: return { ...state, loading: true };
      case fetchSubscriptionDetails.FULFILL: return {
        ...state,
        loading: false,
        loaded: true,
      };

      case SUBSCRIPTION_DETAILS_RECEIVED: return { ...state, ...action.payload };

      case SUBSCRIPTION_DETAILS_PROCESSING: return {
        ...state,
        isSubscriptionDetailsProcessing: action.payload,
      };

      case submitPayment.TRIGGER: return {
        ...state,
        paymentMethod: action.payload.method,
      };
      case submitPayment.REQUEST: return {
        ...state,
        submitting: true,
      };
      case submitPayment.SUCCESS: return {
        ...state,
        redirect: true,
      };
      case submitPayment.FULFILL: return {
        ...state,
        submitting: false,
        paymentMethod: undefined,
      };

      default:
    }
  }
  return state;
};

export default subscriptionDetailsReducer;
