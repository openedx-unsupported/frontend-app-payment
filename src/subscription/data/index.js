export * from './reducers';
export * from './details/actions';
export * from './details/selectors';
export * from './client-secret/actions';
export * from './client-secret/selectors';
export * from './constants';
export * from './service';
export { default as saga } from './sagas';
