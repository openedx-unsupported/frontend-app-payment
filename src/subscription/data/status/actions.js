import { createRoutine } from 'redux-saga-routines';

export const submit3DS = createRoutine('SUBMIT_3DS');

// Actions and their action creators
export const handleSuccessful3DS = createRoutine('HANDLE_SUCCESSFUL_3DS');
export const handleUnSuccessful3DS = createRoutine('HANDLE_UN_SUCCESSFUL_3DS');
export const SUBSCRIPTION_STATUS_RECEIVED = 'SUBSCRIPTION_STATUS_RECEIVED';

export const subscriptionStatusReceived = status => ({
  type: SUBSCRIPTION_STATUS_RECEIVED,
  payload: status,
});
