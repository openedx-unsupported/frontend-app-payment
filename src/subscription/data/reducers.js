import { combineReducers } from 'redux';

import details from './details/reducer';
import clientSecret from './client-secret/reducer';

export const reducer = combineReducers({
  details,
  clientSecret,
});

export default reducer;
