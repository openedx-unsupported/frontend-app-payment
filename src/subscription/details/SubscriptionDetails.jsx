/* eslint-disable max-len */
import React from 'react';
import { useSelector } from 'react-redux';
import {
  FormattedMessage, useIntl,
} from '@edx/frontend-platform/i18n';

import messages from '../../payment/cart/Cart.messages';
import { detailsSelector, currencyDisclaimerSelector } from '../data/details/selectors';

import { CurrencyDisclaimer } from '../../payment/cart/CurrencyDisclaimer';
import OrderSummary from '../../payment/cart/OrderSummary';
import ProductLineItem from '../../payment/cart/ProductLineItem';
import TotalTable from '../../payment/cart/TotalTable';

import SubscriptionDetailsSkeleton from './skeleton/SubscriptionDetailsSkeleton';
import SubscriptionLegal from './legal/SubscriptionLegal';
import SubscriptionContent from './content/SubscriptionContent';
import { SubscriptionSummaryTablePrice } from './summary-table/SubscriptionSummaryTablePrice';
import { SubscriptionOrderDetails } from './order-details/SubscriptionOrderDetails';

/**
 * SubscriptionDetails component
 * renders the subscription details
 */
export const SubscriptionDetails = () => {
  const intl = useIntl();
  const {
    loading,
    loaded,
    products,
    programTitle,
    organizations,
    price,
    programType,
    currency,
    isCurrencyConverted,
    totalPrice,
    isTrialEligible,
    programUuid,
    paymentMethod,
  } = useSelector(detailsSelector);

  const programDetails = {
    programTitle,
    programType,
    organizations,
    price,
  };

  const renderDetails = () => (
    <div>
      <span className="sr-only">
        <FormattedMessage
          id="subscription.screen.reader.details.loaded"
          defaultMessage="Shopping details are loaded."
          description="Screen reader text to be read when details load."
        />
      </span>

      <SubscriptionContent details={programDetails}>
        {products.map(product => (
          <ProductLineItem
            key={product.title}
            {...product}
          />
        ))}
      </SubscriptionContent>
      {loaded
        ? (
          <OrderSummary
            isSubscription
            subscriptionEventProps={{
              isTrialEligible,
              isNewSubscription: isTrialEligible,
              paymentProcessor: paymentMethod,
              programUuid,
            }}
          >
            <SubscriptionSummaryTablePrice
              price={price}
              isTrialEligible={isTrialEligible}
            />
            <TotalTable
              total={totalPrice}
              isSubscription
            />
            {
            isCurrencyConverted
              ? <CurrencyDisclaimer currencyDisclaimerSelector={currencyDisclaimerSelector} />
              : null
            }
          </OrderSummary>
        ) : (
          <>
            <div className="skeleton py-2 mb-3 w-50" />
            <div className="skeleton py-2 mb-2" />
            <div className="skeleton py-2 mb-5" />
          </>
        )}
      <SubscriptionOrderDetails programTitle={programDetails.programTitle} />
      <SubscriptionLegal
        programTitle={programDetails.programTitle}
        price={programDetails.price}
        currency={currency}
      />
    </div>
  );

  return (
    <section
      aria-live="polite"
      aria-relevant="all"
      aria-label={intl.formatMessage(messages['payment.section.cart.label'])}
    >
      {loading ? <SubscriptionDetailsSkeleton /> : renderDetails() }
    </section>
  );
};

export default SubscriptionDetails;
