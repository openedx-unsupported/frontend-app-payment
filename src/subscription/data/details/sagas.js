import {
  call, put, select,
} from 'redux-saga/effects';
import { stopSubmit } from 'redux-form';
import { logError } from '@edx/frontend-platform/logging';

import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { getReduxFormValidationErrors } from '../../../payment/data/utils';

// Actions
import {
  subscriptionDetailsReceived,
  subscriptionDetailsProcessing,
  fetchSubscriptionDetails,
  submitSubscription,
} from './actions';

import { subscriptionStatusReceived } from '../status/actions';

// Sagas
import { handleSubscriptionErrors, clearMessages } from '../../../feedback';

// Services
import * as SubscriptionApiService from '../service';
import { subscriptionStripeCheckout } from '../../subscription-methods';

const sendSubscriptionEvent = ({ details, success }) => {
  const eventType = success
    ? 'edx.bi.user.subscription.program.checkout.success'
    : 'edx.bi.user.subscription.program.checkout.failure';

  sendTrackEvent(
    eventType,
    {
      paymentProcessor: details.paymentMethod,
      isTrialEligible: details.isTrialEligible,
      isNewSubscription: details.isTrialEligible,
      programUuid: details.programUuid,
      price: details.price,
    },
  );
};

export const paymentMethods = {
  stripe: subscriptionStripeCheckout,
};

function* isSubscriptionDetailsProcessing() {
  return yield select(state => state.subscription.details.isSubscriptionDetailsProcessing);
}

export function* handleReduxFormValidationErrors(error) {
  if (error.fieldErrors) {
    const fieldErrors = getReduxFormValidationErrors(error);
    yield put(stopSubmit('subscription', fieldErrors));
  }
}

const handleCustomErrors = (error, fallbackKey) => {
  const apiErrors = [{
    code: fallbackKey ? 'fallback-error' : error.cause,
    userMessage: error.message,
  }];
  const err = new Error();
  err.errors = apiErrors;
  return err;
};

export function* handleFetchSubscriptionDetails() {
  if (yield isSubscriptionDetailsProcessing()) {
    // Do nothing if there is a request currently in flight to get or modify the details
    return;
  }

  try {
    yield put(subscriptionDetailsProcessing(true)); // we are going to modify the details, don't make changes
    const result = yield call(SubscriptionApiService.getDetails);
    yield put(subscriptionDetailsReceived(result));
  } catch (error) {
    const errors = error?.errors || [];
    if (
      errors.length
      && errors[0].code === 'empty_subscription'
    ) {
      yield put(subscriptionDetailsReceived({ errorCode: 'empty_subscription' })); // update redux store with details data
    }
    yield call(handleSubscriptionErrors, error, true);
  } finally {
    yield put(subscriptionDetailsProcessing(false)); // we are done modifying the details
    yield put(fetchSubscriptionDetails.fulfill()); // mark the details as finished loading
  }
}

export function* handleSubmitPayment({ payload }) {
  if (yield isSubscriptionDetailsProcessing()) {
    return;
  }

  const details = yield select(state => ({ ...state.subscription.details }));
  const { method, ...paymentArgs } = payload;
  try {
    yield put(subscriptionDetailsProcessing(true));
    yield put(clearMessages()); // Don't leave messages floating on the page after clicking submit
    yield put(submitSubscription.request());
    const paymentMethodCheckout = paymentMethods[method];
    const postData = yield call(paymentMethodCheckout, details, paymentArgs);
    const result = yield call(SubscriptionApiService.postDetails, postData);

    yield put(submitSubscription.success(result));
    yield put(subscriptionStatusReceived(result));
    // success segment event
    sendSubscriptionEvent({ details, success: true });
  } catch (error) {
    // Do not handle errors on user aborted actions
    if (!error.aborted) {
      if (error.message && error.cause === 'create-paymentMethod') { // stripe payment method creation error
        yield call(handleSubscriptionErrors, handleCustomErrors(error, 'fallback'), true);
      } else {
        logError(error, {
          messagePrefix: 'Stripe-Checkout Post Error',
          paymentMethod: 'Create Subscription',
          paymentErrorType: 'v1/stripe-checkout/ Error',
        });
        yield call(handleSubscriptionErrors, error, true);
      }
    }
    // failure segment event
    sendSubscriptionEvent({ details, success: false });
    yield put(submitSubscription.failure());
  } finally {
    yield put(subscriptionDetailsProcessing(false));
  }
}

export default handleSubmitPayment;
