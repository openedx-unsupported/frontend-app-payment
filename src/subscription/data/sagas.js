import {
  takeLatest,
} from 'redux-saga/effects';

// details
import { fetchSubscriptionDetails, submitSubscription } from './details/actions';
import { handleFetchSubscriptionDetails, handleSubmitSubscription } from './details/sagas';

export default function* saga() {
  yield takeLatest(submitSubscription.TRIGGER, handleSubmitSubscription);
  yield takeLatest(fetchSubscriptionDetails.TRIGGER, handleFetchSubscriptionDetails);
}
