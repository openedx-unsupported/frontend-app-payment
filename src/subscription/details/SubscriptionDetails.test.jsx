/* eslint-disable react/jsx-no-constructed-context-values */
import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import { AppContext } from '@edx/frontend-platform/react';
import { Factory } from 'rosie';
import { createStore } from 'redux';
import Cookies from 'universal-cookie';
import { createSerializer } from 'enzyme-to-json';

import '../__factories__/subscription.factory';
import '../../payment/__factories__/userAccount.factory';
import { SubscriptionDetails } from './SubscriptionDetails';
import createRootReducer from '../../data/reducers';
import { fetchSubscriptionDetails, subscriptionDetailsReceived } from '../data/details/actions';
import { camelCaseObject } from '../../payment/data/utils';

// run enzyme JSON serializer using options compatible with prior snapshots
expect.addSnapshotSerializer(createSerializer({ mode: 'deep', noKey: true }));

jest.mock('@edx/frontend-platform/analytics', () => ({
  sendTrackEvent: jest.fn(),
}));
jest.mock('universal-cookie', () => {
  class MockCookies {
    static result = {};

    get() {
      return MockCookies.result;
    }
  }
  return MockCookies;
});

describe('<SubscriptionDetails />', () => {
  let store;
  let tree;

  beforeEach(() => {
    const authenticatedUser = Factory.build('userAccount');
    Cookies.result = undefined;
    store = createStore(createRootReducer(), {});

    const component = (
      <IntlProvider locale="en">
        <AppContext.Provider value={{ authenticatedUser }}>
          <Provider store={store}>
            <SubscriptionDetails />
          </Provider>
        </AppContext.Provider>
      </IntlProvider>
    );
    tree = mount(component);
  });

  it('renders the loading skeleton', () => {
    expect(tree).toMatchSnapshot();
  });

  it('renders a basic, one product details', () => {
    act(() => {
      store.dispatch(subscriptionDetailsReceived(camelCaseObject(Factory.build(
        'subscription',
        {},
        { numProducts: 1 },
      ))));
      store.dispatch(fetchSubscriptionDetails.fulfill());
    });
    tree.update();
    expect(tree).toMatchSnapshot();
  });

  it('renders 3 product in a subscription details', () => {
    act(() => {
      store.dispatch(subscriptionDetailsReceived(camelCaseObject(Factory.build(
        'subscription',
        {},
        { numProducts: 3 },
      ))));
      store.dispatch(fetchSubscriptionDetails.fulfill());
    });
    tree.update();
    expect(tree).toMatchSnapshot();
  });
});
