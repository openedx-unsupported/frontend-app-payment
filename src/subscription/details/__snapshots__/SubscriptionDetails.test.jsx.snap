// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubscriptionDetails /> renders 3 product in a subscription details 1`] = `
<section
  aria-label="Shopping Cart Details"
  aria-live="polite"
  aria-relevant="all"
>
  <div>
    <span
      className="sr-only"
    >
      Shopping details are loaded.
    </span>
    <div
      className="basket-section"
    >
      <h5
        aria-level="2"
      >
        In your cart
      </h5>
      <div
        className=""
      >
        <div
          className="heading-badge-wrapper"
        >
          <h3
            aria-level="2"
            className="mb-0"
          >
            Blockchain Fundamentals
          </h3>
          <span
            className="badge badge-light"
          >
            Subscription
          </span>
        </div>
        <h4
          aria-level="2"
          className="mb-0"
        >
          Professional Certificate
        </h4>
        <p
          aria-level="2"
          className="body small mb-5"
        >
          University of California, Berkeley
        </p>
        <h5>
          Included with your subscription:
        </h5>
      </div>
      <div
        className="row align-items-center mb-3 product-wrapper"
      >
        <div
          className="col-5"
        >
          <div
            className="embed-responsive embed-responsive-16by9"
          >
            <img
              alt=""
              className="img-thumbnail product-thumbnail"
              src="https://edx.org/watpewal.gif"
            />
          </div>
        </div>
        <div
          className="col-7"
        >
          <h6
            aria-level="3"
            className="m-0"
          >
            Introduction to Pigeonss
          </h6>
          <p
            className="m-0"
          >
            Verified Certificate
          </p>
        </div>
      </div>
      <div
        className="row align-items-center mb-3 product-wrapper"
      >
        <div
          className="col-5"
        >
          <div
            className="embed-responsive embed-responsive-16by9"
          >
            <img
              alt=""
              className="img-thumbnail product-thumbnail"
              src="https://edx.org/lekigud.gif"
            />
          </div>
        </div>
        <div
          className="col-7"
        >
          <h6
            aria-level="3"
            className="m-0"
          >
            Introduction to Pigs
          </h6>
          <p
            className="m-0"
          >
            Verified Certificate
          </p>
        </div>
      </div>
      <div
        className="row align-items-center mb-3 product-wrapper"
      >
        <div
          className="col-5"
        >
          <div
            className="embed-responsive embed-responsive-16by9"
          >
            <img
              alt=""
              className="img-thumbnail product-thumbnail"
              src="https://edx.org/lu.jpg"
            />
          </div>
        </div>
        <div
          className="col-7"
        >
          <h6
            aria-level="3"
            className="m-0"
          >
            Introduction to Guanacos
          </h6>
          <p
            className="m-0"
          >
            Verified Certificate
          </p>
        </div>
      </div>
    </div>
    <div
      aria-labelledby="summary-heading"
      aria-live="polite"
      className="basket-section"
      role="region"
    >
      <h5
        aria-level="2"
        id="summary-heading"
      >
        Summary
      </h5>
      <div
        className="summary-row d-flex"
      >
        <span
          className="flex-grow-1"
        >
          Price
        </span>
        <span
          className="summary-price"
        >
          /month USD after 7-day free trial
        </span>
      </div>
      <div
        className="summary-row font-weight-bold d-flex"
      >
        <span
          className="flex-grow-1"
        >
          Today's total
        </span>
        <span
          className="text-right"
        >
          $49.00
        </span>
      </div>
    </div>
    <div
      className="basket-section order-details"
    >
      <h5
        aria-level="2"
      >
        Order details
      </h5>
      <p
        className="mb-2"
      >
        The above total includes any applicable taxes.
      </p>
      <p
        className="mb-2"
      >
        After you complete your order you will have verified access to each course in Blockchain Fundamentals.
      </p>
      <p>
        Cancel anytime.
      </p>
    </div>
    <p
      className="micro"
    >
      By signing up for a subscription or starting a free trial, you authorize us to charge your card on file $49.00/month for Blockchain Fundamentals and any applicable taxes. If you signed up for a trial, we won’t charge your card until the end of your trial period. Your Blockchain Fundamentals subscription continues until canceled. Monthly subscriptions can be canceled at any time by 
      <a
        className="pgn__hyperlink default-link standalone-link"
        href="https://support.edx.org/hc/en-us/sections/115004173027-Receive-and-Share-edX-Certificates"
        onClick={[Function]}
        target="_self"
      >
        contacting customer support.
      </a>
       Canceling a subscription stops the monthly recurring subscription charge, but doesn't refund the transaction for the current billing period. You can continue to access the subscription until the end of the then current billing period.
    </p>
  </div>
</section>
`;

exports[`<SubscriptionDetails /> renders a basic, one product details 1`] = `
<section
  aria-label="Shopping Cart Details"
  aria-live="polite"
  aria-relevant="all"
>
  <div>
    <span
      className="sr-only"
    >
      Shopping details are loaded.
    </span>
    <div
      className="basket-section"
    >
      <h5
        aria-level="2"
      >
        In your cart
      </h5>
      <div
        className=""
      >
        <div
          className="heading-badge-wrapper"
        >
          <h3
            aria-level="2"
            className="mb-0"
          >
            Blockchain Fundamentals
          </h3>
          <span
            className="badge badge-light"
          >
            Subscription
          </span>
        </div>
        <h4
          aria-level="2"
          className="mb-0"
        >
          Professional Certificate
        </h4>
        <p
          aria-level="2"
          className="body small mb-5"
        >
          University of California, Berkeley
        </p>
        <h5>
          Included with your subscription:
        </h5>
      </div>
      <div
        className="row align-items-center mb-3 product-wrapper"
      >
        <div
          className="col-5"
        >
          <div
            className="embed-responsive embed-responsive-16by9"
          >
            <img
              alt=""
              className="img-thumbnail product-thumbnail"
              src="https://edx.org/vot.png"
            />
          </div>
        </div>
        <div
          className="col-7"
        >
          <h6
            aria-level="3"
            className="m-0"
          >
            Introduction to Llamass
          </h6>
          <p
            className="m-0"
          >
            Verified Certificate
          </p>
        </div>
      </div>
    </div>
    <div
      aria-labelledby="summary-heading"
      aria-live="polite"
      className="basket-section"
      role="region"
    >
      <h5
        aria-level="2"
        id="summary-heading"
      >
        Summary
      </h5>
      <div
        className="summary-row d-flex"
      >
        <span
          className="flex-grow-1"
        >
          Price
        </span>
        <span
          className="summary-price"
        >
          /month USD after 7-day free trial
        </span>
      </div>
      <div
        className="summary-row font-weight-bold d-flex"
      >
        <span
          className="flex-grow-1"
        >
          Today's total
        </span>
        <span
          className="text-right"
        >
          $49.00
        </span>
      </div>
    </div>
    <div
      className="basket-section order-details"
    >
      <h5
        aria-level="2"
      >
        Order details
      </h5>
      <p
        className="mb-2"
      >
        The above total includes any applicable taxes.
      </p>
      <p
        className="mb-2"
      >
        After you complete your order you will have verified access to each course in Blockchain Fundamentals.
      </p>
      <p>
        Cancel anytime.
      </p>
    </div>
    <p
      className="micro"
    >
      By signing up for a subscription or starting a free trial, you authorize us to charge your card on file $49.00/month for Blockchain Fundamentals and any applicable taxes. If you signed up for a trial, we won’t charge your card until the end of your trial period. Your Blockchain Fundamentals subscription continues until canceled. Monthly subscriptions can be canceled at any time by 
      <a
        className="pgn__hyperlink default-link standalone-link"
        href="https://support.edx.org/hc/en-us/sections/115004173027-Receive-and-Share-edX-Certificates"
        onClick={[Function]}
        target="_self"
      >
        contacting customer support.
      </a>
       Canceling a subscription stops the monthly recurring subscription charge, but doesn't refund the transaction for the current billing period. You can continue to access the subscription until the end of the then current billing period.
    </p>
  </div>
</section>
`;

exports[`<SubscriptionDetails /> renders the loading skeleton 1`] = `
<section
  aria-label="Shopping Cart Details"
  aria-live="polite"
  aria-relevant="all"
>
  <div
    className="sub-cart-skeleton"
  >
    <span
      className="sr-only"
    >
      Loading, please wait...
    </span>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton"
        style={
          Object {
            "width": "25%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
    <div
      className="d-flex justify-content-between mb-2 align-items-center"
    >
      <span
        aria-busy={true}
        aria-live="polite"
        className="sub-skeleton-title mr-3"
      >
        <span
          className="react-loading-skeleton"
          style={
            Object {
              "height": "28px",
            }
          }
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy={true}
        aria-live="polite"
        className="sub-skeleton-badge"
      >
        <span
          className="react-loading-skeleton"
          style={
            Object {
              "height": "24px",
            }
          }
        >
          ‌
        </span>
        <br />
      </span>
    </div>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton"
        style={
          Object {
            "width": "45%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton mb-4"
        style={
          Object {
            "width": "65%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
    <div
      className="row align-items-center mb-5"
    >
      <div
        className="col-5"
      >
        <span
          aria-busy={true}
          aria-live="polite"
        >
          <span
            className="react-loading-skeleton mb-2"
            style={
              Object {
                "width": "80%",
              }
            }
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy={true}
          aria-live="polite"
        >
          <span
            className="react-loading-skeleton embed-responsive embed-responsive-16by9"
            style={
              Object {
                "height": "100%",
              }
            }
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        className="col-7 pl-0"
      >
        <span
          aria-busy={true}
          aria-live="polite"
        >
          <span
            className="react-loading-skeleton skeleton mb-2 w-50"
            style={Object {}}
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy={true}
          aria-live="polite"
        >
          <span
            className="react-loading-skeleton skeleton mr-4 py-1"
            style={Object {}}
          >
            ‌
          </span>
          <br />
        </span>
      </div>
    </div>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton"
        style={
          Object {
            "width": "25%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
    <div
      className="d-flex justify-content-between mb-5 price"
    >
      <span
        aria-busy={true}
        aria-live="polite"
        className="section-left"
      >
        <span
          className="react-loading-skeleton"
          style={
            Object {
              "width": "70%",
            }
          }
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy={true}
        aria-live="polite"
        className="section-right"
      >
        <span
          className="react-loading-skeleton"
          style={
            Object {
              "width": "70%",
            }
          }
        >
          ‌
        </span>
        <br />
      </span>
    </div>
    <div
      className="d-flex justify-content-between mb-4 price"
    >
      <span
        aria-busy={true}
        aria-live="polite"
        className="section-left"
      >
        <span
          className="react-loading-skeleton"
          style={
            Object {
              "width": "70%",
            }
          }
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy={true}
        aria-live="polite"
        className="section-right"
      >
        <span
          className="react-loading-skeleton"
          style={
            Object {
              "width": "70%",
            }
          }
        >
          ‌
        </span>
        <br />
      </span>
    </div>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton"
        style={
          Object {
            "width": "25%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton"
        style={
          Object {
            "width": "50%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton mb-4"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton py-5"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        ‌
      </span>
      <br />
    </span>
  </div>
</section>
`;
